packages/jsii/lib/transforms/deprecated-warnings-transformer.ts

```diff
@@ -0,0 +1,413 @@
+
+enum ElementType {
+  METHOD = 0,
+  FUNCTION = 1,
+  CLASS = 2,
+  PARAMETER = 3,
+}
+
+interface Warning {
+  readonly elementName: string;
+  readonly message: string;
+  readonly type: ElementType;
+}
+
+export class DeprecatedWarningInjector {
+  private readonly index = new Map<
+    string,
+    ts.ClassDeclaration | ts.InterfaceDeclaration | undefined
+  >();
+  private deprecatedTypes: ts.Type[] = [];
+  private moduleName = '';
+
+  public constructor(private readonly typeChecker: ts.TypeChecker) {}
+
+  public get customTransformers(): ts.CustomTransformers {
+    return {
+      before: [
+        (context) => {
+          const transformer = new DeprecatedWarningsTransformer(
+            this.typeChecker,
+            context,
+            this.deprecatedTypes,
+            this.index,
+            this.moduleName,
+          );
+          return transformer.transform.bind(transformer);
+        },
+      ],
+    };
+  }
+
+  public process(assembly: Assembly) {
+    if (assembly.types != null) {
+      this.moduleName = assembly.name;
+      const types: Array<InterfaceType | ClassType> = Object.values(
+        assembly.types,
+      ).filter(isClassOrInterfaceType);
+
+      for (const type of types) {
+        const node = bindings.getClassOrInterfaceRelatedNode(type)!;
+        this.index.set(fullyQualifiedName(this.typeChecker, node)!, node);
+      }
+
+      this.deprecatedTypes = Object.values(assembly.types)
+        .filter((typeInfo) => typeInfo.docs?.stability === Stability.Deprecated)
+        .filter(isClassOrInterfaceType)
+        .map((typeInfo) => bindings.getClassOrInterfaceRelatedNode(typeInfo)!)
+        .map((node) => this.typeChecker.getTypeAtLocation(node));
+    }
+  }
+}
+
+class DeprecatedWarningsTransformer {
+  private warnings: Warning[] = [];
+
+  public constructor(
+    private readonly typeChecker: ts.TypeChecker,
+    private readonly context: ts.TransformationContext,
+    private readonly deprecatedTypes: ts.Type[],
+    private readonly index: Map<
+      string,
+      ts.ClassDeclaration | ts.InterfaceDeclaration | undefined
+    >,
+    private readonly moduleName: string,
+  ) {}
+
+  public transform<T extends ts.Node>(node: T): T {
+    if (this.index.size === 0) {
+      // Processing didn't happen, probably due to JSII compilation errors
+      return node;
+    }
+    return this.visitEachChild(node);
+  }
+
+  private visitEachChild<T extends ts.Node>(node: T): T {
+    return ts.visitEachChild(node, this.visitor.bind(this), this.context);
+  }
+
+  private visitor<T extends ts.Node>(node: T): ts.VisitResult<T> {
+    if (ts.isClassDeclaration(node)) {
+      this.handleClassDeclaration(node);
+    }
+
+    if (ts.isFunctionDeclaration(node)) {
+      this.handleFunctionDeclaration(node);
+    }
+
+    if (isMethodLikeDeclaration(node)) {
+      this.handleMethodLikeDeclaration(node);
+    }
+
+    if (ts.isBlock(node)) {
+      if (this.warnings.length > 0) {
```

Comment:
I would not do it like this.

As far as I know (correct me if wrong, but this is what I believe) a "block" is any pair of braces `{ ... }`, so also a pair of braces that could occur within something other than a function declaration like `if (condition) { ... }`.

Right now, it looks like your code assumes that we will do something with warnings in the "first block" that appears after a function declaration? There is some reliance on ordering there that's making me uncomfortable, through mutating the shared state of `this.warnings`.

Can't we just get the block from the method declaration, which we KNOW is the main function body, and work on that directly?

Or am I just misunderstanding how this works?

Reply:
That's a good idea! I had missed the `body` property. However, some reliance on the stated of `this.warnings` is still necessary, I believe; when an entire class is deprecated, for example.